wheatley/bot.py

```diff
@@ -231,31 +258,75 @@ def expect_bell(self, index: int, bell: Bell) -> None:
                 self.stroke
             )
 
-    def start_next_row(self) -> None:
-        """
-        Creates a new row from the row generator and tells the rhythm to expect the new bells.
-        """
-
+    def generate_next_row(self) -> None:
+        """ Creates a new row from the row generator and tells the rhythm to expect the new bells. """
         if self._is_ringing_rounds:
             self._row = self._rounds
         else:
-            self._row = self.row_generator.next_row(self.stroke)
+            self._row, calls = self.row_generator.next_row_and_calls(self.stroke)
+            # Call the calls
+            for c in calls:
+                self._make_call(c)
+            # Add cover bells if needed
             if len(self._row) < len(self._rounds):
-                # Add cover bells if needed
                 self._row.extend(self._rounds[len(self._row):])
 
+    def start_next_row(self, is_first_row: bool) -> None:
+        # Generate the next row and update row indices
+        self._place = 0
+        if is_first_row:
+            self._row_number = 0
+        else:
+            self._row_number += 1
+
+        # Useful local variables
+        has_just_rung_rounds = self._row == self._rounds
+        next_stroke = Stroke.from_index(self._row_number)
+
+        # Implement handbell-style stopping at rounds
+        if self._stop_at_rounds and has_just_rung_rounds and not self._is_ringing_rounds:
+            self._should_stand = False
+            self._is_ringing = False
+
+        # Call any early calls specified by the row generator (if any)
+        if self._rounds_left_before_method is not None:
+            for c in self.row_generator.early_calls().get(self._rounds_left_before_method) or []:
+                self._make_call(c)
+
+        # Start the method if necessary
+        if self._rounds_left_before_method == 0:
+            # Sanity check that we are in fact starting on the correct row (which is no longer
+            # trivially guarunteed)
+            assert next_stroke == self.row_generator.start_stroke()
+            self._rounds_left_before_method = None
+            self._is_ringing_rounds = False
+            # Check that we have the right number of bells, and if so reset the row generator so
+            # that the next row it generates is the first row of the method
+            if self._check_number_of_bells():
+                self.row_generator.reset()
+        # Decrement the counter only if it's not None (here, `None` specficies that we don't know
+        # when the method will start but we should keep ringing rounds
```

Comment:
```suggestion
        # Decrement the counter only if it's not None (here, `None` specifies that we don't know
        # when the method will start but we should keep ringing rounds
```
Also `self._rounds_left_before_method is None` while ringing the method as well as rounds before go is called. 
I think you can remove this comment, I'm happy with what the if statement does from context.

Reply:
Comment deleted.